### DISCUSSION TITLE
Sveltekit Google OAuth2: ReferenceError: EventSource is not defined even though I have EventSource installed and setup globally in vite.config.ts

### DISCUSSION THREAD

**Discussion by LightningFryer** [pocketbase/pocketbase#5869] [Q&A]
So I'm trying to implement Google OAuth2 using Sveltekit and I think I've done everything correctly except whenever I try to hit the OAuth login endpoint using sveltekit actions with this function: 

```
export const actions: Actions = {
    loginGoogle: async ({locals, request}) => {
        const pb = new PocketBase('http://127.0.0.1:8090');
        const authData = await pb.collection('users').authWithOAuth2({ provider: 'google', urlCallback: (url) => {
            window.open(url);
        }});
    }
};
```

(I'm doing window.open(url) here because of a workaround where the endpoint doesn't open the popup, refer to this discussion [here](https://github.com/pocketbase/pocketbase/discussions/2429#discussioncomment-5943061) for that

Here is the frontend which just has a button in a form trying to hit the loginGoogle action endpoint.
```
<script lang="ts">
    import type { PageData } from './$types';
    let { data }: { data: PageData } = $props();
</script>

<form action="?/loginGoogle" method="post">
    <button class="btn" type="submit">Register</button>
</form>
```

When the button is clicked I get this error:

```
ClientResponseError 0: Something went wrong while processing your request.
    at file:///C:/Users/User/Desktop/VSC%20temp/Github%20Repos/graphica-core/node_modules/pocketbase/dist/pocketbase.es.mjs:1:23653
    at process.processTicksAndRejections (node:internal/process/task_queues:105:5) {
  url: '',
  status: 0,
  response: {},
  isAbort: false,
  originalError: ReferenceError: EventSource is not defined
      at RealtimeService.initConnect (file:///C:/Users/User/Desktop/VSC%20temp/Github%20Repos/graphica-core/node_modules/pocketbase/dist/pocketbase.es.mjs:1:17218)
      at file:///C:/Users/User/Desktop/VSC%20temp/Github%20Repos/graphica-core/node_modules/pocketbase/dist/pocketbase.es.mjs:1:16971
      at new Promise (<anonymous>)
      at RealtimeService.connect (file:///C:/Users/User/Desktop/VSC%20temp/Github%20Repos/graphica-core/node_modules/pocketbase/dist/pocketbase.es.mjs:1:16866)
      at RealtimeService.subscribe (file:///C:/Users/User/Desktop/VSC%20temp/Github%20Repos/graphica-core/node_modules/pocketbase/dist/pocketbase.es.mjs:1:14261)
      at file:///C:/Users/User/Desktop/VSC%20temp/Github%20Repos/graphica-core/node_modules/pocketbase/dist/pocketbase.es.mjs:1:22919
      at new Promise (<anonymous>)
      at file:///C:/Users/User/Desktop/VSC%20temp/Github%20Repos/graphica-core/node_modules/pocketbase/dist/pocketbase.es.mjs:1:22881
      at process.processTicksAndRejections (node:internal/process/task_queues:105:5)
      at async loginGoogle (eval at instantiateModule (file:///C:/Users/User/Desktop/VSC%20temp/Github%20Repos/graphica-core/node_modules/vite/dist/node/chunks/dep-BWSbWtLw.js:52963:24), <anonymous>:12:22)
}
```

I have eventsource installed with npm and I declared it globally in vite.config.ts like this. I have not used eventsource in the past so I don't know much about it so if this is where I'm wrong then please tell me how to fix it:
```
import { sveltekit } from '@sveltejs/kit/vite';
import { defineConfig } from 'vite';
import EventSource from 'eventsource';

export default defineConfig({
	plugins: [sveltekit()],
	define: {
		global: {
			EventSource: EventSource,
		}
	}
});
```

Please tell me what I'm doing wrong. Any help is much appreciated!

**Comment by orenaksakal** [comment]


**Reply by ganigeorgiev** [reply]
The `// good for safari` comment is wrong. As mentioned in the linked discussion Safari doesn't allow for some reason opening windows in async handlers and the above most likely will fail. The JS SDK already handles this out-of-the box but the main point is that the call shouldn't use `await` and the click handler shouldn't use `async` (here is an example from my other open source project - https://github.com/presentator/presentator/blob/09609c49c67275bf42ec7769d59a9d3d72f63645/ui/src/components/base/SocialAuth.svelte#L9).

**Comment by ganigeorgiev** [comment]
@LightningFryer Please don't open the same question in both the issues tracker and the discussions.

Couple notes:

- The linked workound for Safari mentioned in https://github.com/pocketbase/pocketbase/discussions/2429#discussioncomment-5943061 is no longer needed because the JS SDK does it by default but in order to work your handler must not be async (_additionally if you ignore that this is fixed in the recent JS SDK you are still not doing the same thing as in the suggestion answer because the window must be opened BEFORE the async handler_).

- as others already mentioned, `authWithOAuth2()` is expected to run in the browser. From the above code sample it seems that you are trying to execute the call in a server action and this will not work because it needs to be able to open a browser window and maintain a realtime connection for the duration of the OAuth2 auth. If you want a server only handler you need to use the "Manual code exchange flow", aka. the second tab in https://pocketbase.io/docs/authentication/#oauth2-integration (the example is using plain JS and localstorage but the idea is the same with the only difference that instead of localStorage you'll have to use cookies).

In any case, there are at least dozens other similar related posts.

If you are new to web development I strongly recommend against using JS SSR with PocketBase because it comes with a lot of complications and you'll probably mess up the security and I don't know how nor I have the time to explain it better. 

**Please make sure to read https://github.com/pocketbase/pocketbase/discussions/5313 and more specifically "Problem 1" and "Problem 2".**
